sql/sql_table.cc

```diff
@@ -11746,6 +11746,60 @@ bool mysql_trans_commit_alter_copy_data(THD *thd)
   DBUG_RETURN(error);
 }
 
+/** Handle the error when copying data from source to target table.
+@param error          error code
+@param ignore         alter ignore statement
+@param to             target table handler
+@param thd            Mysql Thread
+@param alter_ctx      Runtime context for alter statement
+@return true in case of error
+@return false in case of skipping the row and continue alter operation */
+static bool
+copy_data_error_handler(int &error, bool ignore, TABLE *to,
+                        THD *thd, Alter_table_ctx *alter_ctx)
+{
+  if (to->file->is_fatal_error(error, HA_CHECK_DUP))
+  {
+    /* Not a duplicate key error. */
+    to->file->print_error(error, MYF(0));
+    error = 1;
+func_exit:
+    return true;
+  }
+  /* Duplicate key error. */
+  if (unlikely(alter_ctx->fk_error_if_delete_row))
+  {
+    /* We are trying to omit a row from the table which serves
+    as parent in a foreign key. This might have broken
+    referential integrity so emit an error. Note that we
+    can't ignore this error even if we are
+    executing ALTER IGNORE TABLE. IGNORE allows to skip rows, but
+    doesn't allow to break unique or foreign key constraints, */
```

Comment:
Is the `ALTER IGNORE TABLE` scenario covered by some test? That test case name had better be mentioned in the commit message. (Mentioning it in a source code comment is a worse idea, because someone could rename or refactor the test and forget to adjust that comment. In the commit message, the comment would naturally refer to the current state of affairs.)

Reply:
In a6bb0ca6ca2ff12a1110a2bf9c6f5bab84c57f7b this is covered by the addition to the test `innodb.foreign_key`.